File: app/src/main/java/com/example/administrator/itemdecoration/itemdecoration/OverlayDecoration.java
package com.example.administrator.itemdecoration.itemdecoration;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.LayoutParams;
import android.support.v7.widget.RecyclerView.State;
import android.view.View;

/**
 * Created by lengwei on 2017/11/6.
 *
 * @Description 使用onDrawOver绘制分割线
 */
public class OverlayDecoration extends RecyclerView.ItemDecoration {

    private static final int [] ATTRS = new int[]{android.R.attr.listDivider};
    private Context mContext;
    private int orientation = LinearLayoutManager.VERTICAL;
    private Drawable mDrawable;


    private OverlayDecoration(Context context) {
        this.mContext = context;
        final TypedArray a = context.obtainStyledAttributes(ATTRS);
        this.mDrawable = a.getDrawable(0);
        a.recycle();
    }

    public OverlayDecoration(Context context, int orientation) {
        this(context);
        this.orientation = orientation;
    }

    @Override
    public void onDrawOver(Canvas c, RecyclerView parent, State state) {
        super.onDrawOver(c, parent, state);
        if (orientation == LinearLayoutManager.VERTICAL) {
            drawVertical(c, parent);
        } else {
            drawHorizontal(c, parent);
        }
    }

    private void drawVertical(Canvas c, RecyclerView parent){
        final int left = parent.getPaddingLeft();
        final int right = parent.getWidth() - parent.getPaddingRight();
        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            View child = parent.getChildAt(i);
            if (parent.getChildAdapterPosition(child) == parent.getAdapter().getItemCount() - 1){
                continue;
            }

            final RecyclerView.LayoutParams lp = (LayoutParams) child.getLayoutParams();
            int bottom = child.getBottom() + lp.bottomMargin + Math.round(child.getTranslationY());
            int top = bottom - mDrawable.getIntrinsicHeight();
            mDrawable.setBounds(left, top, right, bottom);
            mDrawable.draw(c);
        }
    }

    private void drawHorizontal(Canvas c, RecyclerView parent){
        final int top = parent.getPaddingTop();
        final int bottom = parent.getHeight() - parent.getPaddingBottom();
        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            View child = parent.getChildAt(i);
            if (parent.getChildAdapterPosition(child) == parent.getAdapter().getItemCount() - 1){
                return;
            }

            final RecyclerView.LayoutParams lp = (LayoutParams) child.getLayoutParams();
            int right = child.getRight() + lp.rightMargin + Math.round(child.getTranslationY());
            int left = right - mDrawable.getIntrinsicHeight();
            mDrawable.setBounds(left, top, right, bottom);
            mDrawable.draw(c);
        }
    }
}

File: app/src/main/java/com/example/administrator/itemdecoration/itemdecoration/OverlayDecoration.kt
package com.example.administrator.itemdecoration.itemdecoration

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.LayoutParams
import android.support.v7.widget.RecyclerView.State

/**
 * Created by lengwei on 2017/11/6.
 *
 * @Description 使用onDrawOver绘制分割线
 */
class OverlayDecoration private constructor(context: Context) : RecyclerView.ItemDecoration() {
    private var orientation = LinearLayoutManager.VERTICAL
    private val mDrawable: Drawable?


    init {
        val a = context.obtainStyledAttributes(ATTRS)
        this.mDrawable = a.getDrawable(0)
        a.recycle()
    }

    constructor(context: Context, orientation: Int) : this(context) {
        this.orientation = orientation
    }

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: State) {
        super.onDrawOver(c, parent, state)
        if (orientation == LinearLayoutManager.VERTICAL) {
            drawVertical(c, parent)
        } else {
            drawHorizontal(c, parent)
        }
    }

    private fun drawVertical(c: Canvas, parent: RecyclerView) {
        val left = parent.paddingLeft
        val right = parent.width - parent.paddingRight
        val childCount = parent.childCount
        for (i in 0 until childCount) {
            val child = parent.getChildAt(i)
            val adapter = parent.adapter ?: continue
            if (parent.getChildAdapterPosition(child) == adapter.itemCount - 1) {
                continue
            }

            val lp = child.layoutParams as LayoutParams
            val bottom = child.bottom + lp.bottomMargin + Math.round(child.translationY)
            val top = bottom - mDrawable!!.intrinsicHeight
            mDrawable.setBounds(left, top, right, bottom)
            mDrawable.draw(c)
        }
    }

    private fun drawHorizontal(c: Canvas, parent: RecyclerView) {
        val top = parent.paddingTop
        val bottom = parent.height - parent.paddingBottom
        val childCount = parent.childCount
        for (i in 0 until childCount) {
            val child = parent.getChildAt(i)
            val adapter = parent.adapter ?: continue
            if (parent.getChildAdapterPosition(child) == adapter.itemCount - 1) {
                return
            }

            val lp = child.layoutParams as LayoutParams
            val right = child.right + lp.rightMargin + Math.round(child.translationY)
            val left = right - mDrawable!!.intrinsicHeight
            mDrawable.setBounds(left, top, right, bottom)
            mDrawable.draw(c)
        }
    }

    companion object {
        private val ATTRS = intArrayOf(android.R.attr.listDivider)
    }
}
